File: frontend/src/lib/api/manyfold.ts
import { apiFetch } from './core'

// ---------------------------------------------------------------------------
// Manyfold connector — Part 3: admin API client
//
// Mirrors backend/app/routers/manyfold.py. Multiple self-hosted Manyfold
// instances can be registered by an admin; each has an OAuth2
// (client_credentials) client_id/client_secret pair. The secret is
// write-only — POST/PATCH accept it, but no response ever echoes it back.
// `has_secret` tells the UI whether one is stored.
// ---------------------------------------------------------------------------

export interface ManyfoldInstance {
  id: number
  base_url: string
  domain: string
  display_name: string | null
  client_id: string
  /** Whether a client_secret is stored server-side — the secret itself is never returned. */
  has_secret: boolean
  scopes: string
  enabled: boolean
  last_connected_at: string | null
  notes: string | null
  created_at: string
  updated_at: string
}

export interface CreateManyfoldInstanceRequest {
  base_url: string
  display_name?: string | null
  client_id: string
  /** Plaintext; encrypted server-side before storage. Write-only. */
  client_secret: string
  scopes?: string
  enabled?: boolean
  notes?: string | null
}

export interface PatchManyfoldInstanceRequest {
  base_url?: string | null
  display_name?: string | null
  client_id?: string | null
  /** If provided, rotates the stored secret. Write-only — omit to keep the existing one. */
  client_secret?: string | null
  scopes?: string | null
  enabled?: boolean | null
  notes?: string | null
}

export interface ManyfoldTestConnectionResult {
  ok: boolean
  scope?: string | null
  message?: string | null
}

export const listManyfoldInstances = (): Promise<ManyfoldInstance[]> =>
  apiFetch<ManyfoldInstance[]>('/api/admin/manyfold')

export const getManyfoldInstance = (id: number): Promise<ManyfoldInstance> =>
  apiFetch<ManyfoldInstance>(`/api/admin/manyfold/${id}`)

export const createManyfoldInstance = (
  body: CreateManyfoldInstanceRequest,
): Promise<ManyfoldInstance> =>
  apiFetch<ManyfoldInstance>('/api/admin/manyfold', {
    method: 'POST',
    body: JSON.stringify(body),
  })

export const patchManyfoldInstance = (
  id: number,
  body: PatchManyfoldInstanceRequest,
): Promise<ManyfoldInstance> =>
  apiFetch<ManyfoldInstance>(`/api/admin/manyfold/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(body),
  })

export const deleteManyfoldInstance = (id: number): Promise<void> =>
  apiFetch<void>(`/api/admin/manyfold/${id}`, { method: 'DELETE' })

export const testManyfoldConnection = (
  id: number,
): Promise<ManyfoldTestConnectionResult> =>
  apiFetch<ManyfoldTestConnectionResult>(
    `/api/admin/manyfold/${id}/test-connection`,
    { method: 'POST', body: '{}' },
  )

File: backend/app/models/manyfold.py
"""ManyfoldInstance — admin-configured Manyfold connector instances (Part 1 of 3).

Manyfold is a self-hosted 3D-model organizer with an OAuth2 (client_credentials)
API. An admin registers one or more Manyfold instances by domain, pasting an
OAuth client ID + client secret. Later parts (Part 2: connector/worker/download,
Part 3: frontend) use these credentials to import a model straight from a
Manyfold URL.

``base_url`` is the full origin used for API calls (e.g.
``https://manyfold.crzynet.com``). ``domain`` is the host-only form (e.g.
``manyfold.crzynet.com``), derived from ``base_url`` and kept unique so Part 2
can match an import URL's domain to the right instance.

``client_secret_enc`` is Fernet-encrypted (via ``app.crypto``) — never stored
in plaintext, never returned by the API. See docs/decisions.md.
"""

from datetime import datetime

from sqlalchemy import Boolean, DateTime, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from .base import Base


class ManyfoldInstance(Base):
    __tablename__ = "manyfold_instances"

    id: Mapped[int] = mapped_column(primary_key=True)

    # Full origin used for API calls, e.g. "https://manyfold.crzynet.com".
    # Normalized on write: trailing slash stripped, host lowercased,
    # scheme restricted to http/https.
    base_url: Mapped[str] = mapped_column(String(512), nullable=False)

    # Host-only form, e.g. "manyfold.crzynet.com". Derived from base_url.
    # Unique — Part 2 matches an import URL's domain against this column.
    domain: Mapped[str] = mapped_column(
        String(255), nullable=False, unique=True, index=True
    )

    display_name: Mapped[str | None] = mapped_column(String(255), nullable=True)

    # OAuth client_credentials identifier — not a secret, stored plaintext.
    client_id: Mapped[str] = mapped_column(String(255), nullable=False)

    # Fernet-encrypted OAuth client secret. Nullable so a row can exist
    # pre-secret (e.g. mid-setup), but the admin router requires it on create.
    client_secret_enc: Mapped[str | None] = mapped_column(Text, nullable=True)

    # OAuth scope string requested on token fetch. "public" alone runs
    # anonymous; "read" is needed to see owner-private models.
    scopes: Mapped[str] = mapped_column(
        String(255), nullable=False, default="public read"
    )

    enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    # Set on a successful POST /{id}/test-connection call.
    last_connected_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True
    )

    notes: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    def __repr__(self) -> str:
        return f"<ManyfoldInstance id={self.id} domain={self.domain!r}>"
